qiskit_experiments/framework/backend_timing.py

```diff
@@ -0,0 +1,363 @@
+            timing = BackendTiming(self.backend)
+
+            circs = []
+            for duration in self.experiment_options.durations:
+                # Calculate valid sample number closest to this duration
+                circ = template_circ.assign_parameters(
+                    {dur: timing.pulse_samples(duration)},
+                    inplace=False,
+                )
+                # Track corresponding duration in seconds for the pulse
+                circ.metadata = {
+                    "xval": timing.pulse_time(duration),
+                    "unit": "s",
+                }
+    """
+
+    def __init__(self, backend: Backend):
+        """Initialize backend timing object
+
+        Args:
+            experiment: the experiment to provide timing help for
+        """
+        backend_data = BackendData(backend)
+
+        # Pull all the timing data from the backend
+        self._acquire_alignment = backend_data.acquire_alignment
+        self._granularity = backend_data.granularity
+        self._min_length = backend_data.min_length
+        self._pulse_alignment = backend_data.pulse_alignment
+        #: The backend's ``dt`` value, copied to :class:`.BackendTiming` for convenience
+        self.dt = backend_data.dt
+
+    @property
+    def delay_unit(self) -> str:
+        """The delay unit for the current backend
+
+        "dt" is used if dt is present in the backend configuration. Otherwise
+        "s" is used.
+        """
+        if self.dt is not None:
+            return "dt"
+
+        return "s"
+
+    def circuit_delay(self, time: float) -> Union[int, float]:
+        """Delay duration closest to ``time`` and consistent with timing constraints
+
+        This method produces the value to pass for the ``duration`` of a
+        ``Delay`` instruction of a ``QuantumCircuit`` so that the delay fills
+        the time until the next valid pulse, assuming the ``Delay`` instruction
+        begins on a sample that is also valid for a pulse to begin on.
+
+        The pulse timing constraints of the backend are considered in order to
+        give the number of samples closest to ``time`` for the start of a pulse
+        in a subsequent instruction to be valid. The least common multiple of
+        the pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        If :meth:`.BackendTiming.delay_unit` is ``s``, ``time`` is
+        returned directly. Typically, this is the case for a simulator where
+        converting to sample number is not needed.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples if :meth:`.BackendTiming.delay_unit`
+            is ``dt``. Otherwise return ``time``.
+        """
+        if self.delay_unit == "s":
+            return time
+
+        return self.schedule_delay(time)
+
+    def schedule_delay(self, time: float) -> int:
```

Comment:
Hmm, that is the correct interpretation -- `circuit_delay` is the delay in a circuit program and `schedule_delay` is the delay in a pulse schedule. Why does it not seem like that is the expected usage? Is it that `circuit_delay` calls `schedule_delay` that seems wrong? They need the same logic, but I could make a hidden method that they both call so that they appear to be on more equal footing instead of `circuit_delay` depending on `schedule_delay`.

Grammatically, I agree the names are not the best. They are more appropriate as property names than method names since method names are usually verbs. The most accurate would be something like `calculate_circuit_delay` but that is long. Would `delay_circuit` / `delay_schedule` be better? Or maybe `calc_circuit_delay` or `get_circuit_delay`?

Reply:
Methods refactored in [d531cc1](https://github.com/Qiskit/qiskit-experiments/pull/860/commits/d531cc1ed4f85136153096f090c4e041d1f8ec5b).